### PR TITLE
[Feature][Connector][GoogleFirestore-Sink] Support GoogleFirestore Sink

### PR DESCRIPTION
**Pull request by lvshaokang**
## Purpose of this pull request

Support google firestore sink.

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/incubator-seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Comment by TaoZex** [comment]
Please provide a screenshot or e2e test to prove that it is available.

**Comment by lvshaokang** [comment]
@TaoZex
I didn't find any available docker image, so I don't know if it is still possible to do e2e testing.
The following is a screenshot of the manual testing.
image

**Comment by TaoZex** [comment]
LGTM. Thanks for your contribution.

File: seatunnel-connectors-v2/connector-google-firestore/pom.xml [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+        <dependency>
+            <groupId>org.apache.seatunnel</groupId>
+            <artifactId>connector-common</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>com.google.cloud</groupId>
+            <artifactId>google-cloud-firestore</artifactId>
+            <version>3.7.10</version>
```
Comment (TyrantLucifer) [inline seatunnel-connectors-v2/connector-google-firestore/pom.xml]: Use properties to manage version number.

File: seatunnel-connectors-v2/connector-google-firestore/pom.xml [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+            <version>31.1-android</version>
```
Comment (TyrantLucifer) [inline seatunnel-connectors-v2/connector-google-firestore/pom.xml]: The same as above.

File: seatunnel-connectors-v2/connector-google-firestore/src/main/java/org/apache/seatunnel/connectors/seatunnel/google/firestore/sink/FirestoreSinkWriter.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        collectionReference.add(serializer.serialize(seaTunnelRow));
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (firestore != null) {
+            try {
+                firestore.close();
+            } catch (Exception e) {
+                throw new RuntimeException(e);
```
Comment (TyrantLucifer) [inline seatunnel-connectors-v2/connector-google-firestore/src/main/java/org/apache/seatunnel/connectors/seatunnel/google/firestore/sink/FirestoreSinkWriter.java]: Unified exception

File: seatunnel-connectors-v2/connector-google-firestore/pom.xml (hunk 1 above)
Comment (lvshaokang) [reply seatunnel-connectors-v2/connector-google-firestore/pom.xml]: done

**Comment by lvshaokang** [comment]
@TyrantLucifer PTAL. Thanks!

**Comment by liugddx** [comment]
you also should update release-note https://github.com/apache/incubator-seatunnel/blob/dev/release-note.md

**Comment by liugddx** [comment]
please add e2e test case. Refer to https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/coding-guide.md

**Comment by lvshaokang** [comment]
> @TaoZex I didn't find any available docker image, so I don't know if it is still possible to do e2e testing. The following is a screenshot of the manual testing. image

refer

File: seatunnel-connectors-v2/connector-google-firestore/src/main/java/org/apache/seatunnel/connectors/seatunnel/google/firestore/exception/FirestoreConnectorErrorCode.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.seatunnel.connectors.seatunnel.google.firestore.exception;
+
+import org.apache.seatunnel.common.exception.SeaTunnelErrorCode;
+
+public enum FirestoreConnectorErrorCode implements SeaTunnelErrorCode {
+    CLOSE_CLIENT_FAILED("FIRESTORE-01", "Close Firestore client failed");
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-google-firestore/src/main/java/org/apache/seatunnel/connectors/seatunnel/google/firestore/exception/FirestoreConnectorErrorCode.java:23]: Please add this error code to `docs/en/connector-v2/Error-Quick-Reference-Manual.md`

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-google-firestore-e2e/src/test/resources/firestore/fake_to_google_firestore.conf [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        c_bytes = bytes
+        c_date = date
+        c_timestamp = timestamp
+      }
+    }
+    result_table_name = "fake"
+  }
+}
+
+transform {
```
Comment (EricJoy2048) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-google-firestore-e2e/src/test/resources/firestore/fake_to_google_firestore.conf]: Remove this transform if it is unnecessary.

**Comment by EricJoy2048** [comment]
I have no problem, @liugddx  @TaoZex  @TyrantLucifer  PTAL again.

File: seatunnel-connectors-v2/connector-google-firestore/src/main/java/org/apache/seatunnel/connectors/seatunnel/google/firestore/sink/FirestoreSinkWriter.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+    @Override
+    public void close() throws IOException {
+        if (firestore != null) {
+            try {
+                firestore.close();
+            } catch (Exception e) {
+                log.error("Close Firestore client failed.", e);
+                throw new FirestoreConnectorException(
+                        FirestoreConnectorErrorCode.CLOSE_CLIENT_FAILED, e);
```
Comment (TaoZex) [inline seatunnel-connectors-v2/connector-google-firestore/src/main/java/org/apache/seatunnel/connectors/seatunnel/google/firestore/sink/FirestoreSinkWriter.java]:
```suggestion
                throw new FirestoreConnectorException(
                        FirestoreConnectorErrorCode.CLOSE_CLIENT_FAILED, "Close Firestore client failed.", e);
```
Logging and throwing exception can cause log duplication.

**Review by TyrantLucifer — DISMISSED** [review]

**Comment by TyrantLucifer** [comment]
LGTM, thank you for your contribution.

**Review by TaoZex — DISMISSED** [review]
LGTM

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-google-firestore-e2e/src/test/java/org.apache.seatunnel.e2e.connector.google.firestore/GoogleFirestoreIT.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    @TestTemplate
+    public void testGoogleFirestore(TestContainer container) throws Exception {
+        Container.ExecResult execResult = container.executeJob(FIRESTORE_CONF_FILE);
+        Assertions.assertEquals(0, execResult.getExitCode());
+
+        List<QueryDocumentSnapshot> documents = readSinkDataset();
+        Assertions.assertTrue(documents.size() > 1);
+        Assertions.assertEquals(15, documents.get(0).getData().size());
```
Comment (ic4y) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-google-firestore-e2e/src/test/java/org.apache.seatunnel.e2e.connector.google.firestore/GoogleFirestoreIT.java]: Refer to src/test/java/org/apache/seatunnel/e2e/connector/starrocks/StarRocksCDCSinkIT.java, it is necessary to check the value of each field in each row

Comment (lvshaokang) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-google-firestore-e2e/src/test/java/org.apache.seatunnel.e2e.connector.google.firestore/GoogleFirestoreIT.java]: @ic4y PTAL. Thanks

**Review by TaoZex — APPROVED** [review]

**Review by TyrantLucifer — APPROVED** [review]
LGTM, cc @hailin0 @EricJoy2048

**Review by ic4y — APPROVED** [review]

**Comment by ic4y** [comment]
@lvshaokang  Suggest adding the [new version documentation](https://github.com/apache/incubator-seatunnel/issues/4544) for GoogleFirestore Sink in subsequent PR.Or create a related issue

**Comment by lvshaokang** [comment]
> @lvshaokang Suggest adding the [new version documentation](https://github.com/apache/incubator-seatunnel/issues/4544) for GoogleFirestore Sink in subsequent PR.Or create a related issue

@ic4y ok, i will supply new doc follow the rules.

**Pull request merged**